libs/src/service-selection/ServiceSelection.js

```diff
@@ -0,0 +1,223 @@
+    this.backups = {}
+
+    // Total number of services attempted
+    this.totalAttempts = 0
+  }
+
+  async select () {
+    // If a short circuit is provided, take it. Don't check it, just use it.
+    const shortcircuit = this.shortcircuit()
+    if (shortcircuit) return shortcircuit
+
+    // Get all the services
+    let services = await this.getServices()
+
+    // If a whitelist is provided, filter down to it
+    if (this.whitelist) {
+      services = this.filterToWhitelist(services)
+    }
+
+    // Filter out anything we know is already unhealthy
+    const filteredServices = this.filterOutKnownUnhealthy(services)
+
+    // Randomly sample a "round" to test
+    const round = this.getSelectionRound(filteredServices)
+
+    this.totalAttempts += round.length
+
+    // If there are no services left to try, either pick a backup or return null
+    if (filteredServices.length === 0) {
+      if (Object.keys(this.backups).length > 0) {
+        // Some backup exists
+        return this.selectFromBackups()
+      } else {
+        // Nothing could be found that was healthy.
+        // Reset everything we know so that we might try again.
+        this.unhealthy = []
+        this.backups = {}
+        return null
+      }
+    }
+
+    // Race this "round" of services, getting the best and ones that errored
+    // Note: ones that did not error or were not the best just get canceled so
+    // we don't really know anything about them at this point.
+    const { best, errored } = await this.race(round)
+
+    // Mark all the errored ones as unhealthy
+    errored.forEach(e => this.addUnhealthy(e))
+
+    // Trigger a cleanup event for all of the unhealthy and backup services,
+    // so they can get retried in the future
+    this.triggerCleanup()
+
+    // Recursively try this selection function if we didn't find something
+    if (!best) {
+      return this.select()
+    }
+
+    // If we made it this far, we found the best service! (of the rounds we tried)
+    return best
+  }
+
+  /** Triggers a clean up of unhealthy and backup services so they can be retried later */
+  triggerCleanup () {
+    clearTimeout(this.unhealthyCleanupTimeout)
+    clearTimeout(this.backupCleanupTimeout)
+
+    this.unhealthyCleanupTimeout = setTimeout(() => { this.unhealthy = [] }, this.unhealthyTTL)
+    this.backupCleanupTimeout = setTimeout(() => { this.backups = {} }, this.backupsTTL)
+  }
+
+  /** A short-circuit. If overriden, can be used to skip selection (which could be slow) */
+  shortcircuit () {
+    return null
+  }
+
+  /** Filter down services to those in the whitelist */
+  filterToWhitelist (services) {
+    return services.filter(s => this.whitelist.has(s))
+  }
+
+  /** Filter out known unhealthy services from the provided */
+  filterOutKnownUnhealthy (services) {
+    return services.filter(s => !this.unhealthy.has(s))
+  }
+
+  /** Given a list of services, samples maxConcurrentRequests from them */
+  getSelectionRound (services) {
```

Comment:
why is this sampling necessary? since we already have a list. seems like we might end up eliminating valid endpoints if the value for 'this.maxConcurrentRequests' is < services.length.

Reply:
unless that is by design

Reply:
So it just does multiple rounds if nothing in this "round" works. The reason for having this check is to make sure the browser isn't getting hammered (chrome supports max 6 concurrent requests).